### PR TITLE
add PipelineError to capture exit code

### PR DESCRIPTION
**Pull request by simonbyrne**
At the moment if a process fails and throws and error, there is no easy way to capture the exit code. The only other option is to run asynchronously and manually check (but then there is no easy way to throw an error...)

### CODE REVIEW

**Comment by simonbyrne** [comment]
Alternative to #26719

File: base/download.jl [hunk 1]
```diff
@@ -11,7 +11,17 @@ if Sys.iswindows()
...
+        try
+            run(`$ps -Version 3 -NoProfile -Command "$tls12; $downloadfile"`)
+        catch e
+            if e isa PipelineError{Process} && e.pipe.exitcode % Int32 == -393216
+                # appears to be "wrong version" exit code, based on
+                # https://docs.microsoft.com/en-us/azure/cloud-services/cloud-services-startup-tasks-common
+                error("Downloading files requires Windows Management Framework 3.0 or later.")
+            end
+            rethrow(e)
+        end
```
Comment (JeffBezanson) [inline base/download.jl]: This should probably be a separate commit.

Comment (simonbyrne) [reply base/download.jl]: Same PR okay?

Comment (JeffBezanson) [reply base/download.jl]: Sure.

**Comment by StefanKarpinski** [comment]
Some bikeshedding: this could be called `ProcessError` since there's always some process that failed, and it's not always wrapping a pipeline. Maybe this structure:
```jl
struct ProcessError <: Exception
    process::Process
    pipeline::Union{Nothing,Pipeline}
end
```
Where the `process` is the process that failed and `pipeline` is the pipeline that it belonged to, if any.

**Comment by simonbyrne** [comment]
> @vtjnash doesn't want to use it for this particular use case, favoring a run(..., wait=false) call with a status check instead.

So similar to #26719?

**Comment by simonbyrne** [comment]
Can't you have multiple failures in the one pipeline?

File: base/download.jl (hunk 1 above)
Comment (simonbyrne) [reply base/download.jl]: In that case, should we export `pipeline_error` (or some sort of "throw error based on exit code" function?)

**Comment by StefanKarpinski** [comment]
> Can't you have multiple failures in the one pipeline?

Sure, that's why it has the pipeline field as well.

**Comment by StefanKarpinski** [comment]
The first one that failed, presumably.

**Comment by JeffBezanson** [comment]
This can still be done but doesn't block 1.0.

**Comment by simonbyrne** [comment]
Looks like it was fixed by #27900
